ethereum/core/src/main/java/org/hyperledger/besu/ethereum/chain/ChainDataPruner.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright Hyperledger Besu Contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+package org.hyperledger.besu.ethereum.chain;
+
+import org.hyperledger.besu.datatypes.Hash;
+import org.hyperledger.besu.plugin.services.storage.KeyValueStorageTransaction;
+
+import java.util.Collection;
+import java.util.concurrent.ExecutorService;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ChainDataPruner implements BlockAddedObserver {
+  public static final int MAX_PRUNING_WORKER = 16;
+  private static final Logger LOG = LoggerFactory.getLogger(ChainDataPruner.class);
+  private final BlockchainStorage blockchainStorage;
+  private final ChainDataPrunerStorage prunerStorage;
+  private final long blocksToRetain;
+  private final long pruningFrequency;
+  private final ExecutorService pruningExecutor;
+
+  public ChainDataPruner(
+      final BlockchainStorage blockchainStorage,
+      final ChainDataPrunerStorage prunerStorage,
+      final long blocksToRetain,
+      final long pruningFrequency,
+      final ExecutorService pruningExecutor) {
+    this.blockchainStorage = blockchainStorage;
+    this.prunerStorage = prunerStorage;
+    this.blocksToRetain = blocksToRetain;
+    this.pruningFrequency = pruningFrequency;
+    this.pruningExecutor = pruningExecutor;
+  }
+
+  @Override
+  public void onBlockAdded(final BlockAddedEvent event) {
+    final long blockNumber = event.getBlock().getHeader().getNumber();
+    final long storedPruningMark = prunerStorage.getPruningMark().orElse(blockNumber);
+    if (blockNumber < storedPruningMark) {
+      LOG.warn(
+          "Block added event: "
+              + event
+              + " has a block number of "
+              + blockNumber
+              + " < pruning mark "
+              + storedPruningMark
+              + " which normally indicates the blocksToRetain is too small");
+      return;
+    }
+    final KeyValueStorageTransaction tx1 = prunerStorage.startTransaction();
+    final Collection<Hash> forkBlocks = prunerStorage.getForkBlocks(blockNumber);
+    forkBlocks.add(event.getBlock().getHash());
+    prunerStorage.setForkBlocks(tx1, blockNumber, forkBlocks);
+    tx1.commit();
+
+    pruningExecutor.submit(
```

Comment:
Can another thread update the forkBlocks in the DB in between these transactions?
If so, then could that be a problem? 

...I'm thinking not because it's pulled out of storage again in the method below, but would like you to confirm it's been considered :)

Reply:
Yes, it is possible but it won't be a problem. Additions to forkBlocks are happening in order and the pruning thread is to be created after any additions to forkBlocks. Therefore, when a pruning operation is happening in a thread for a given blockNumber, we can be sure that no additions would be performed on forkBlocks < blockNumber, only deletion will be performed. Deletions don't need to happen in order, so this should be fine.

Reply:
> we can be sure that no additions would be performed on forkBlocks < blockNumber
might need to chat to get my head around this...I'm wondering how the case where we might import an 'old' block might be handled, e.g. we built on a fork then had a reorg and CL had to tell us about old blocks